src/iotHubMessageExplorer.ts

```diff
@@ -21,27 +69,103 @@ export class IoTHubMessageExplorer extends IoTHubMessageBaseExplorer {
+    }
+
+    private async delay(ms: number, token?: vscode.CancellationToken) {
+        if (ms <= 1000) {
+            return new Promise( resolve => setTimeout(resolve, ms));
+        } else {
+            await new Promise( resolve => setTimeout(resolve, 1000));
+            if (token) {
+                if (token.isCancellationRequested) {
+                    return;
+                } else {
+                    await this.delay(ms - 1000, token);
+                }
+            } else {
+                await this.delay(ms - 1000);
+            }
+        }
+    }
+
+    public async sendD2CMessageFromMultipleDevicesRepeatedlyWithProgressBar(deviceConnectionStrings: string[], message: string, times: number, interval: number) {
+        const startTime = new Date();
+        this.outputLine(Constants.SimulatorSummaryLabel, `Start at ${startTime}`)
+        await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification,
+			title: Constants.SimulatorSendingMessageProgressBarTitle,
+			cancellable: true
+		}, async (progress, token) => {
+            token.onCancellationRequested(() => {
+                vscode.window.showInformationMessage(Constants.SimulatorProgressBarCancelLog);
+            })
+            progress.report({ increment: 0});
+            const deviceCount = deviceConnectionStrings.length;
+            const total = deviceCount * times;
+            if (total <= 0) {
+                this.outputLine(Constants.SimulatorSummaryLabel, `Invalid Operation.`);
+                return;
+            }
+            const step = 100 / total;
+            let clients = [];
+            let statuses = [];
+            let ids = [];
+            let totalStatus = new SendStatus('Total', total);
+            for(let i = 0; i < deviceCount; i++) {
+                clients.push(await clientFromConnectionString(deviceConnectionStrings[i]));
+                statuses.push(new SendStatus(ConnectionString.parse(deviceConnectionStrings[i]).DeviceId, times));
+                ids.push(i);
+            }
+            const sendingStartTime = new Date();
+            for(let i = 0; i < times; i++) {
+                await ids.map(async j => await this.sendD2CMessageCoreWithProgress(clients[j], message, statuses[j], totalStatus));
+                progress.report({
+                    increment: step * deviceCount,
+                    message: `Sending message(s) ${totalStatus.sum()} of ${total}`
+                })
+                if (token.isCancellationRequested) {
+                    break;
+                }
+                await this.delay(interval, token);
+            }
+            const sendingEndTime = new Date();
+            this.outputLine(Constants.SimulatorSummaryLabel, `Sending ${total} message(s) done in ${(sendingEndTime.getTime() - sendingStartTime.getTime()) / 1000} second(s), please wait a few seconds for the result.`);

```

Comment:
Calculate the progress based on how many messages finished sending, instead of how many messages started sending